### PR TITLE
fix: [ci] release workflow non-functional

### PR DESCRIPTION
**Pull request by spencerdrak**
Release workflow was misconfigured and had been failing since Friday - we did not get any notifications about this because the workflow file was invalid and could not run (therefore, it couldn't send us the slack message about it's failure). In addition, we've changed the name of the OSX artifacts which would've broken, had this workflow been able to run successfully.

This PR gets release back on the rails by:
- allowing jobs that build wheels to pass an artifact name as output, to standardize usage
- Adding the required missing job for aarch64 linux wheels. 

Test Plan:
- release dry-run here: https://github.com/returntocorp/semgrep/actions/runs/5582933738

I'll re-enable release workflow once this goes in.

@cgdolan it looks like a while back, you set up linting on PRs for GHA that I would've expected to catch this - I'm going to dig into that separately.

PR checklist:

- [ ] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Review by cgdolan — APPROVED** [review]

**Review by tpetr — APPROVED** [review]
Thank you! Sorry for the headache here

File: .github/workflows/build-test-osx-arm64.yaml
Comment (aryx) [inline .github/workflows/build-test-osx-arm64.yaml]: Is there anywhere where we set the wheel-artifact-name for this workflow to something other than osx-arm64-wheel?
If not, then what's the point of introducing this extra intermediate constant that looks like a variable but is not?

Comment (spencerdrak) [reply .github/workflows/build-test-osx-arm64.yaml]: In release.yml, there is one other place this is referenced in the `upload-wheels` job. Unfortunately, we can't set an env var in release.yml to hold this value, but this will likely be helpful with some follow up work I have on deck to factorize the OSX workflows.

Comment (aryx) [reply .github/workflows/build-test-osx-arm64.yaml]: if it's referenced but never set to a different value, it means it's a constant and we should just inline its content instead of introducing those intermediate name that complicate things IMHO.

Comment (spencerdrak) [reply .github/workflows/build-test-osx-arm64.yaml]: Ok, I can remove that bit then. thanks!

Comment (spencerdrak) [reply .github/workflows/build-test-osx-arm64.yaml]: @aryx swapped that around to make it such that the wheel artifact names are outputs which can be defined once and referenced, not inputs - lmk if that's acceptable. thanks!

File: .github/workflows/build-test-manylinux-aarch64.yaml
Comment (aryx) [inline .github/workflows/build-test-manylinux-aarch64.yaml]: Honestly the way we do things in Yaml looks completely foreign to me. If we were in a sane programming language,
we could just return values, takes parameters, define constant in a shared file, and we would not need all those weird ways to do very simple things.

File: .github/workflows/release.yml
Comment (aryx) [inline .github/workflows/release.yml]: This solution looks even worse to me. But again why we introduce those intermediate step variables in the first place?

File: .github/workflows/release.yml
Comment (aryx) [inline .github/workflows/release.yml]: Why not simply replace this name: ${{ ...} with the actual value we know it contains? It's a constant!
What's the gain of using this very long variable name, than even contain this ...manylinux-x86-wheel.... in it
with simply manylinux-x86-wheel!
It's a shorter, and it removes the need for those ENV:, for those needs.wait.xxx everywhere, for those echo ... >> $GITHUB_OUTPUT.

**Review by aryx — DISMISSED** [review]
Looks too complicated. Just use directly 'manylinux-x86-wheel' where it's needed.

**Comment by cgdolan** [comment]
> Looks too complicated. Just use directly 'manylinux-x86-wheel' where it's needed.

Directly using these string values in multiple places is the problem we're trying to solve in this PR. We realized that the next release would likely fail after the changes in https://github.com/returntocorp/semgrep/pull/8286 because the wheel producer step outputs an artifact under a different name than the wheel consumer step expects. The fix for this is to declare the output name once and reference it in such a way that if the value changes all callers get the updated value, which is what we've implemented here.

**Review by cgdolan — APPROVED** [review]

**Comment by aryx** [comment]
And how using steps.xxx.whateverlong.manywheel-x86-arm variable names will be better than using the manywheel-x86-arm string?
What do you win? If you mispelled something, the string or the variable name you will get a runtime error in both cases, so I'd rather use the simpler code.

**Comment by cgdolan** [comment]
With variables, invalid references are surfaced much earlier before/at PR time by IDEs and workflow linting whereas with raw strings invalid references are surfaced much later at workflow runtime.

Here's what I see when I introduce a typo for example:

Screenshot 2023-07-18 at 8 30 07 AM

**Comment by aryx** [comment]
Does the linting work across files? If you reference an output from a workflow in another yaml file is it detected?
If you make a typo in echo "manylinux-x86-wheel-artifact-name
is it detected?

**Comment by spencerdrak** [comment]
Kicked off a release dry-run with these new changes here: https://github.com/returntocorp/semgrep/actions/runs/5589925058

**Comment by cgdolan** [comment]
In the interest of getting this PR submitted and unblocking the release process, I've downscoped the changes to a minimal subset needed to fix the blockers. @aryx PTAL.

**Review by tpetr — APPROVED** [review]
LGTM assuming the dry run succeeds

**Comment by cgdolan** [comment]
Dry run has succeeded and I've implemented the initial feedback as asked. I'm going to dismiss the review requesting changes so we can get this in and unblock releases.

**Review by aryx — COMMENTED** [review]
Thx chris. Appreciate.

**Pull request merged**
